web/src/narrow.js

```diff
@@ -123,6 +122,32 @@ export function activate(raw_terms, opts) {
          or rerendering due to server-side changes.
     */
 
+    // The empty narrow is the All messages view.
+    if (raw_terms.length === 0) {
+        raw_terms = [{operator: "is", operand: "home"}];
+    }
+    const filter = new Filter(raw_terms);
+
+    const is_narrowed_to_all_messages_view = narrow_state.filter()?.is_in_home();
+    if (has_visited_all_messages && is_narrowed_to_all_messages_view && filter.is_in_home()) {
+        // If we're already looking at the All messages view, exit without doing any work.
+        return;
+    }
+
+    if (filter.is_in_home() && message_scroll_state.actively_scrolling) {
+        // `All messages` narrow.
+        // TODO: Figure out why puppeteer test for this fails when run for narrows
+        // other than `All messages`.
+
+        // There is no way to intercept in-flight scroll events, and they will
+        // cause you to end up in the wrong place if you are actively scrolling
+        // on an unnarrow. Wait a bit and try again once the scrolling is likely over.
+        setTimeout(() => {
+            activate(raw_terms, opts);
+        }, 50);
+        return;
+    }
+
     // Use to determine if user read any unread messages in non-All Messages narrow.
     const was_narrowed_already = narrow_state.filter() !== undefined;
```

Comment:
This still returns the same value as before, since `narrow_state.filter()` still returns `undefined` for "all messages".